### PR TITLE
Added documentation and where clause for API `uniqued()`

### PR DESCRIPTION
**Pull request by Roshankumar350**
Problem: 
    * A quick look up for the API named `uniqued()` with example  associated with it was missing.
    * API named `uniqued()` while enumerating each element was checking an `if` condition.

Solution:
    * API documentation added to address the same.
    *  `where` clause added for the same

### Checklist
- [ yes] I've added at least one test that validates that my change is working, if appropriate
- [ yes] I've followed the code style of the rest of the project
- [ yes] I've read the [Contribution Guidelines](../CONTRIBUTING.md)
- [ yes] I've updated the documentation if necessary

### CODE REVIEW

**Comment by xwu** [comment]
Good catch on the typo!

>  I've followed the code style of the rest of the project

Note that the code style involves indentations with _two_ spaces, and also wrapping to 80 columns on each line.

File: Sources/Algorithms/Unique.swift [hunk 1]
```diff
@@ -14,33 +14,57 @@
 //===----------------------------------------------------------------------===//
 
 extension Sequence where Element: Hashable {
-  /// Returns an array with only the unique elements of this sequence, in the
-  /// order of the first occurence of each unique element.
+  /// Returns an array with only  unique elements of this sequence, in the
+  /// order of the first occurrence of each unique element.
+  ///
+  /// In this example, `uniqued()` is used to include only names which are
+  /// unique
```
Comment (xwu) [inline Sources/Algorithms/Unique.swift:21]: Thanks, the formatting is now mostly consistent. Make sure to include periods at the end of sentences, like here (also elsewhere.) There is still another closing brace that is incorrectly indented. There are also several places where you have two consecutive spaces in the comment or the code; please make sure to remove them. There are also places where there are missing spaces after punctuation: please make sure to add them.

A more general comment here: I don’t think it adds anything to describe the example here by saying that `uniqued` makes the array unique.

File: Sources/Algorithms/Unique.swift [hunk 2]
```diff
@@ -14,33 +14,57 @@
...
+  /// by the given projection), in the order of the first occurrence of each
   /// unique element.
+  ///
+  /// In this example, `uniqued()` is used to include only names which are
+  /// unique
+  ///
+  ///     let animals = ["dog", "pig", "cat", "ox", "dog","cat"]
+  ///     let uniqued = animals.uniqued()
+  ///     print(uniqued)
+  ///     // Prints "["dog", "pig","cat", "ox"]"`
```
Comment (natecook1000) [inline Sources/Algorithms/Unique.swift:52]: Could you switch this to an example that demonstrates the `uniqued(on:)` method? This example shows the parameterless method defined above.

File: Sources/Algorithms/Unique.swift [hunk 3]
```diff
@@ -14,33 +14,57 @@
...
+  ///
+  /// - Parameter projection: A projecting closure. `projection` accepts an
+  ///   element of this sequence as its parameter and returns a projected
+  ///   value of the same  type.
+  ///
+  /// - Returns: Returns an array with only the unique elements of this sequence
```
Comment (natecook1000) [inline Sources/Algorithms/Unique.swift:58]:
```suggestion
  /// - Returns: An array with only the unique elements of this sequence, as
  ///   determined by the result of `projection` for each element.
```

File: Sources/Algorithms/Unique.swift [hunk 4]
```diff
@@ -14,33 +14,57 @@
...
+  ///
+  /// In this example, `uniqued()` is used to include only names which are
+  /// unique
+  ///
+  ///     let animals = ["dog", "pig", "cat", "ox", "dog","cat"]
+  ///     let uniqued = animals.uniqued()
+  ///     print(uniqued)
+  ///     // Prints "["dog", "pig","cat", "ox"]"`
+  ///
+  /// - Returns: Returns an array with only the unique elements of this sequence
```
Comment (natecook1000) [inline Sources/Algorithms/Unique.swift:28]:
```suggestion
  /// - Returns: An array with only the unique elements of this sequence.
```

File: Sources/Algorithms/Unique.swift [hunk 5]
```diff
@@ -14,33 +14,57 @@
...
+  ///  .
+  /// - Complexity: O(*n*), where *n* is the length of the sequence.
   public func uniqued() -> [Element] {
     var seen: Set<Element> = []
     var result: [Element] = []
-    for element in self {
-      if seen.insert(element).inserted {
+    for element in self where seen.insert(element).inserted {
```
Comment (natecook1000) [inline Sources/Algorithms/Unique.swift:34]: We don't need this kind of lateral change, here and below.

**Review by natecook1000 — CHANGES_REQUESTED** [review]
Thanks for filling out this documentation, @Roshankumar350! A few notes below for you.

File: Sources/Algorithms/Unique.swift [hunk 6]
```diff
@@ -30,8 +39,22 @@ extension Sequence where Element: Hashable {
 
 extension Sequence {
   /// Returns an array with the unique elements of this sequence (as determined
-  /// by the given projection), in the order of the first occurence of each
+  /// by the given projection), in the order of the first occurrence of each
   /// unique element.
+  ///
+  ///     let animals = ["dog", "pig", "cat", "ox", "dog","cat"]
+  ///     let uniqued = animals(on: { $0.first })
```
Comment (xwu) [inline Sources/Algorithms/Unique.swift:46]: This does not compile.

File: Sources/Algorithms/Unique.swift [hunk 7]
```diff
@@ -30,8 +39,22 @@ extension Sequence where Element: Hashable {
...
+  ///     print(uniqued)
+  ///     // Prints "["dog", "pig","cat", "ox"]"
```
Comment (xwu) [inline Sources/Algorithms/Unique.swift:48]:
```suggestion
  ///     // Prints '["dog", "pig", "cat", "ox"]'
```

Although the output depends on what the projection chosen for this example is.

File: Sources/Algorithms/Unique.swift [hunk 8]
```diff
@@ -30,8 +39,22 @@ extension Sequence where Element: Hashable {
...
+  ///
+  /// - Parameter projection: A projecting closure. `projection` accepts an
+  ///   element of this sequence as its parameter and returns a projected
+  ///   value of the same  type.
```
Comment (xwu) [inline Sources/Algorithms/Unique.swift:52]:
```suggestion
  ///   value of the same type.
```

This description is true, but it’s also just describing the type signature in words. But what are the semantics expected of this closure? That should be documented here.

Comment (Roshankumar350) [reply Sources/Algorithms/Unique.swift]: Added constraint of "Hashable" element

File: Sources/Algorithms/Unique.swift (hunk 6 above)
Comment (Roshankumar350) [reply Sources/Algorithms/Unique.swift]: Addressed and tested the same example on compiler.

File: Sources/Algorithms/Unique.swift [hunk 9]
```diff
@@ -42,14 +42,14 @@ extension Sequence {
   /// by the given projection), in the order of the first occurrence of each
   /// unique element.
   ///
-  ///     let animals = ["dog", "pig", "cat", "ox", "dog","cat"]
-  ///     let uniqued = animals(on: { $0.first })
+  ///     let animals = ["dog", "pig", "cat", "ox", "dog", "cat"]
```
Comment (xwu) [inline Sources/Algorithms/Unique.swift:45]: This example doesn't really demonstrate the role of the 'projection' closure. If you want to stick to the projection `$0.first`, I'd suggest something like:

```suggestion
  ///     let animals = ["dog", "pig", "cat", "ox", "cow", "owl"]
```

File: Sources/Algorithms/Unique.swift [hunk 10]
```diff
@@ -42,14 +42,14 @@ extension Sequence {
...
+  ///     let uniqued = animals.uniqued(on: {$0.first})
   ///     print(uniqued)
-  ///     // Prints "["dog", "pig","cat", "ox"]"
+  ///     // Prints '["dog", "pig", "cat", "ox"]'
   ///
   /// - Parameter projection: A projecting closure. `projection` accepts an
   ///   element of this sequence as its parameter and returns a projected
-  ///   value of the same  type.
+  ///   value of the same type having constraint `Hashable`.
```
Comment (xwu) [inline Sources/Algorithms/Unique.swift:52]: This is not correct: the projected value does *not* need to be of the same type.

But moreover, even when corrected, this statement is still just describing the type of the parameter in words, which is not necessary and doesn't add anything. The challenge here is to teach the user why and how to use a projection function to determine uniqueness.

File: Sources/Algorithms/Unique.swift [hunk 11]
```diff
@@ -30,8 +39,23 @@ extension Sequence where Element: Hashable {
...
+  ///
+  ///     let animals = ["dog", "pig", "cat", "ox", "cow", "owl"]
+  ///     let uniqued = animals.uniqued(on: {$0.first})
+  ///     print(uniqued)
+  ///     // Prints '["dog", "pig", "cat", "ox"]'
+  ///
+  /// - Parameter projection: A projecting closure. `projection` accepts an
+  ///   element of this sequence as its parameter which is having the type of
+  ///   projecting element and returns a projected value that may have the same
+  ///   type having constraint `Hashable`.
```
Comment (xwu) [inline Sources/Algorithms/Unique.swift:53]: @natecook1000 Do you have suggestions of how to approach describing this parameter?

Comment (natecook1000) [inline Sources/Algorithms/Unique.swift:53]: How about this?

```suggestion
  /// - Parameter projection: A closure that transforms an element into the
  ///   value to use for uniqueness. If `projection` returns the same value
  ///   for two different elements, the second element will be excluded
  ///   from the resulting array.
```

File: Sources/Algorithms/Unique.swift [hunk 12]
```diff
@@ -30,8 +39,23 @@ extension Sequence where Element: Hashable {
 
 extension Sequence {
   /// Returns an array with the unique elements of this sequence (as determined
-  /// by the given projection), in the order of the first occurence of each
+  /// by the given projection), in the order of the first occurrence of each
   /// unique element.
+  ///
+  ///     let animals = ["dog", "pig", "cat", "ox", "cow", "owl"]
+  ///     let uniqued = animals.uniqued(on: {$0.first})
```
Comment (natecook1000) [inline Sources/Algorithms/Unique.swift:46]: Let's give an introduction to this example, as well:

```suggestion
  /// This example finds the elements of the `animals` array with unique
  /// first characters:
  ///
  ///     let animals = ["dog", "pig", "cat", "ox", "cow", "owl"]
  ///     let uniqued = animals.uniqued(on: { $0.first })
```

**Review by natecook1000 — COMMENTED** [review]
Getting close, @Roshankumar350!

File: Sources/Algorithms/Unique.swift (hunk 11 above)
Comment (Roshankumar350) [reply Sources/Algorithms/Unique.swift:53]: @natecook1000 Thanks for forward thinking.

**Comment by xwu** [comment]
This looks great!

Final step: looks like there's a conflict between this change and the main branch, and the revisions here are best as one commit instead of 18. So please squash and rebase this PR. Here's an introductory guide to doing that from the folks at Servo:

https://github.com/servo/servo/wiki/Beginner%27s-guide-to-rebasing-and-squashing

File: Sources/Algorithms/Unique.swift [hunk 13]
```diff
@@ -15,18 +15,50 @@
...
+    return result
   }
 }
 
 extension Sequence {
   /// Returns an array with the unique elements of this sequence (as determined
-  /// by the given projection), in the order of the first occurence of each
+  /// by the given projection), in the order of the first occurrence of each
   /// unique element.
-  @inlinable
```
Comment (xwu) [inline Sources/Algorithms/Unique.swift:29]: I don’t think this was rebased correctly. These annotations shouldn’t be deleted.

**Comment by natecook1000** [comment]
Looks great, @Roshankumar350 — thanks for sticking with this! 🎉

**Pull request merged**